mlir/lib/Dialect/Affine/Analysis/Utils.cpp

```diff
@@ -129,13 +133,24 @@ bool MemRefDependenceGraph::init() {
       for (auto *opInst : collector.loadOpInsts) {
         node.loads.push_back(opInst);
         auto memref = cast<AffineReadOpInterface>(opInst).getMemRef();
+        if (memrefAccesses.find(memref) == memrefAccesses.end()) {
+          for (auto *n : memoryEffectNodes)
+            memrefAccesses[memref].insert(n->id);
+        }
         memrefAccesses[memref].insert(node.id);
       }
       for (auto *opInst : collector.storeOpInsts) {
         node.stores.push_back(opInst);
         auto memref = cast<AffineWriteOpInterface>(opInst).getMemRef();
+        if (memrefAccesses.find(memref) == memrefAccesses.end()) {
+          for (auto *n : memoryEffectNodes)
+            memrefAccesses[memref].insert(n->id);
+        }
         memrefAccesses[memref].insert(node.id);
       }
+      node.memEffects = collector.memoryEffectOpInsts;
```

Comment:
If the current for-node exhibits memory effects, it is appended to the sequential list of 'for-nodes' that have memory effects.